ext/mbstring/mbstring.c

```diff
@@ -3390,28 +3390,20 @@ static zend_string* html_numeric_entity_decode(zend_string *input, const mbfl_en
 				uint32_t w = *++p2;
 				while ((w >= '0' && w <= '9') || (w >= 'A' && w <= 'F') || (w >= 'a' && w <= 'f'))
 					w = *++p2;
-				if (p2 == wchar_buf + out_len) {
-					/* We hit the end of the buffer while still reading digits */
-					if ((p2 - p) <= 11 && in_len) {
-						/* The number of digits was legal (1-8 hex digits)
-						 * We need to process this identity again on the next iteration of the main loop */
-						memmove(wchar_buf, p, (p2 - p) * 4);
-						wchar_buf_offset = p2 - p;
-						goto process_converted_wchars;
-					} else {
-						/* Invalid entity */
-						memcpy(converted, p, (p2 - p) * 4);
-						converted += p2 - p;
-						goto process_converted_wchars_no_offset;
-					}
-				} else if ((p2 - p) == 3 || (p2 - p) > 11) {
-					/* Invalid entity */
+				if ((p2 == wchar_buf + out_len) && in_len && (p2 - p) <= 11) {
```

Comment:
Please note... the reason for the `<= 11` check is to avoid getting stuck in an infinite loop if we have a huge numeric entity which fills the entire wchar buffer.

Probably I should make a unit test for that (bizarre) case...

Reply:
Unit test is added.

Reply:
Is this `11` actually `(3 * (sizeof *wchar_buf) + 1)`. *Maybe* make that explicit?

Reply:
`p2` and `p` are both pointers to `uint32_t`, so 11 is the count of 32-bit values between them, not 8-bit values.

11 is the length of `&#x` plus 8 hexadecimal digits, which is the maximum number which can appear in a valid hexadecimal entity.

Reply:
Oh, of course! Not sure if it's worth to spell that out (`sizeof "&#x" - 1 + 8`  doesn't read particularly good).

Reply:
I could make it a macro...

Reply:
I like that. Thank you!